File: legacy-python/data_structures/prediction_market.py
import random
from dataclasses import dataclass, field
from typing import List, Tuple, Optional


@dataclass
class Prediction:
    question: str
    resolve_step: int
    target: Optional[object] = None
    bets: List[Tuple[object, str, float]] = field(default_factory=list)
    resolved: bool = False
    outcome: Optional[str] = None


class PredictionMarket:
    def __init__(self, dao, treasury, event_bus=None) -> None:
        self.dao = dao
        self.treasury = treasury
        self.event_bus = event_bus
        self.predictions: List[Prediction] = []

    def create_prediction(self, question: str, resolve_step: int, target: object | None = None) -> Prediction:
        prediction = Prediction(question, resolve_step, target)
        self.predictions.append(prediction)
        if self.event_bus:
            self.event_bus.publish(
                "prediction_created",
                step=self.dao.current_step,
                question=question,
                resolve_step=resolve_step,
            )
        return prediction

    def place_bet(self, member, prediction: Prediction, choice: str, amount: float) -> bool:
        if amount <= 0 or member.tokens < amount or prediction.resolved:
            return False
        member.tokens -= amount
        self.treasury.deposit("DAO_TOKEN", amount, step=self.dao.current_step)
        prediction.bets.append((member, choice, amount))
        if self.event_bus:
            self.event_bus.publish(
                "bet_placed",
                step=self.dao.current_step,
                member=member.unique_id,
                question=prediction.question,
                choice=choice,
                amount=amount,
            )
        return True

    def _determine_outcome(self, prediction: Prediction) -> str:
        if prediction.target and hasattr(prediction.target, "status"):
            return "pass" if getattr(prediction.target, "status") == "approved" else "fail"
        return random.choice(["pass", "fail"])

    def resolve_predictions(self, step: int) -> None:
        remaining: List[Prediction] = []
        for pred in self.predictions:
            if pred.resolved or step < pred.resolve_step:
                remaining.append(pred)
                continue
            outcome = self._determine_outcome(pred)
            pred.outcome = outcome
            pred.resolved = True
            winners = [(m, amt) for m, choice, amt in pred.bets if choice == outcome]
            total_pool = sum(amt for _, _, amt in pred.bets)
            total_winning = sum(amt for _, amt in winners)
            if total_winning:
                for member, amt in winners:
                    share = total_pool * (amt / total_winning)
                    gained = self.treasury.withdraw("DAO_TOKEN", share, step=step)
                    member.tokens += gained
            if self.event_bus:
                self.event_bus.publish(
                    "prediction_resolved",
                    step=step,
                    question=pred.question,
                    outcome=outcome,
                )
        self.predictions = remaining

File: lib/data-structures/prediction-market.ts
// Prediction Market - betting system on proposal outcomes
// Port from data_structures/prediction_market.py

import type { DAO } from './dao';
import type { Treasury } from './treasury';
import type { EventBus } from '../utils/event-bus';
import type { DAOMember } from '../agents/base';
import { randomBool } from '../utils/random';

export interface Prediction {
  question: string;
  resolveStep: number;
  target: unknown | null;
  bets: Array<[DAOMember, string, number]>;
  resolved: boolean;
  outcome: string | null;
}

export class PredictionMarket {
  dao: DAO;
  treasury: Treasury;
  eventBus: EventBus | null;
  predictions: Prediction[] = [];
  private resolvedPredictions: Prediction[] = [];

  constructor(dao: DAO, treasury: Treasury, eventBus?: EventBus) {
    this.dao = dao;
    this.treasury = treasury;
    this.eventBus = eventBus || null;
  }

  /**
   * Create a new prediction
   */
  createPrediction(
    question: string,
    resolveStep: number,
    target: unknown = null
  ): Prediction {
    const prediction: Prediction = {
      question,
      resolveStep,
      target,
      bets: [],
      resolved: false,
      outcome: null,
    };

    this.predictions.push(prediction);

    if (this.eventBus) {
      this.eventBus.publish('prediction_created', {
        step: this.dao.currentStep || 0,
        question,
        resolveStep,
      });
    }

    return prediction;
  }

  /**
   * Place a bet on a prediction
   */
  placeBet(
    member: DAOMember,
    prediction: Prediction,
    choice: string,
    amount: number
  ): boolean {
    if (
      !this.predictions.includes(prediction) ||
      prediction.resolved ||
      (choice !== 'pass' && choice !== 'fail') ||
      !Number.isFinite(amount) ||
      amount <= 0 ||
      !Number.isFinite(member.tokens) ||
      member.tokens < amount
    ) {
      return false;
    }

    // Deduct tokens from member
    member.tokens -= amount;

    // Deposit into treasury
    this.treasury.deposit('DAO_TOKEN', amount, this.dao.currentStep || 0);

    // Record bet
    prediction.bets.push([member, choice, amount]);

    if (this.eventBus) {
      this.eventBus.publish('bet_placed', {
        step: this.dao.currentStep || 0,
        member: member.uniqueId,
        question: prediction.question,
        choice,
        amount,
      });
    }

    return true;
  }

  /**
   * Determine the outcome of a prediction
   */
  private determineOutcome(prediction: Prediction): string {
    // If target has status property, use it
    if (typeof prediction.target === 'object' && prediction.target !== null) {
      const status = (prediction.target as { status?: unknown }).status;
      if (typeof status === 'string') {
        return status === 'approved' ? 'pass' : 'fail';
      }
    }

    // Otherwise random
    return randomBool() ? 'pass' : 'fail';
  }

  /**
   * Resolve all predictions that are due
   */
  resolvePredictions(step: number): void {
    const remaining: Prediction[] = [];

    for (const pred of this.predictions) {
      if (pred.resolved || step < pred.resolveStep) {
        remaining.push(pred);
        continue;
      }

      // Resolve this prediction
      const outcome = this.determineOutcome(pred);
      pred.outcome = outcome;
      pred.resolved = true;

      // Calculate winners
      const winners = pred.bets.filter(([, choice]) => choice === outcome);
      const totalPool = pred.bets.reduce((sum, [, , amt]) => sum + amt, 0);
      const totalWinning = winners.reduce((sum, [, , amt]) => sum + amt, 0);

      if (totalWinning > 0) {
        // Distribute winnings proportionally
        for (const [member, , amt] of winners) {
          const share = totalPool * (amt / totalWinning);
          const gained = this.treasury.withdraw('DAO_TOKEN', share, step);
          member.tokens += gained;
        }
      }

      if (this.eventBus) {
        this.eventBus.publish('prediction_resolved', {
          step,
          question: pred.question,
          outcome,
        });
      }

      this.resolvedPredictions.push(pred);
    }

    this.predictions = remaining;
  }

  /**
   * Get all active predictions
   */
  getActivePredictions(): Prediction[] {
    return this.predictions.filter((p) => !p.resolved);
  }

  /**
   * Get resolved predictions retained for learning and analytics.
   */
  getResolvedPredictions(): Prediction[] {
    return [...this.resolvedPredictions];
  }

  /**
   * Get prediction statistics
   */
  getStatistics(): {
    total: number;
    active: number;
    resolved: number;
    totalBets: number;
    totalVolume: number;
  } {
    const allPredictions = [...this.predictions, ...this.resolvedPredictions];
    const active = this.predictions.filter((p) => !p.resolved).length;
    const resolved = this.resolvedPredictions.length;
    const totalBets = allPredictions.reduce(
      (sum, p) => sum + p.bets.length,
      0
    );
    const totalVolume = allPredictions.reduce(
      (sum, p) => sum + p.bets.reduce((s, [, , amt]) => s + amt, 0),
      0
    );

    return {
      total: allPredictions.length,
      active,
      resolved,
      totalBets,
      totalVolume,
    };
  }
}
